src/rgw/sync/metadata.cc

```diff
@@ -0,0 +1,309 @@
+                 LogStatus& log_status)
+    -> awaitable<void>
+{
+  auto prefix = PeriodPrefix{*dpp, period};
+  dpp = &prefix;
+
+  co_await ceph::async::parallel_for_each(
+      std::views::iota(0u, shards), // XXX: iota is broken on clang
+      [&] (uint32_t shard) -> awaitable<void> {
+        co_await sync_shard(dpp, period, shard, peer_log, local_log,
+                            peer_meta, local_meta, log_status);
+      });
+}
+
+auto sync_shard_locked(const DoutPrefixProvider* dpp,
+                       std::string_view period, uint32_t shard,
+                       RemoteLog& peer_log, LocalLog& local_log,
+                       RemoteMetadata& peer_meta, LocalMetadata& local_meta,
+                       LogStatus& log_status)
+    -> awaitable<void>
+{
+  RGWObjVersionTracker objv;
+  rgw_meta_sync_marker shard_status;
+  co_await log_status.read(shard, objv, shard_status);
+
+  if (shard_status.state == rgw_meta_sync_marker::SyncState::FullSync) {
+  }
+
+  if (shard_status.state == rgw_meta_sync_marker::SyncState::IncrementalSync) {
+  }
+}
+
+// satisfies LockClient with calls to LogStatus::lock/unlock
+struct ShardLockClient : LockClient {
+  LogStatus& status;
+  uint32_t shard;
+  std::string_view cookie;
+
+  ShardLockClient(LogStatus& status, uint32_t shard, std::string_view cookie)
+    : status(status), shard(shard), cookie(cookie) {}
+
+  awaitable<void> acquire(ceph::timespan duration) override {
+    constexpr bool renew = false;
+    co_await status.lock(shard, cookie, duration, renew);
+  }
+  awaitable<void> renew(ceph::timespan duration) override {
+    constexpr bool renew = true;
+    co_await status.lock(shard, cookie, duration, renew);
+  }
+  awaitable<void> release() override {
+    co_await status.unlock(shard, cookie);
+  }
+};
+
+class ShardPrefix : public DoutPrefixPipe {
+  const uint32_t shard;
+ public:
+  ShardPrefix(const DoutPrefixProvider& dpp, uint32_t shard)
+    : DoutPrefixPipe(dpp), shard(shard) {}
+  void add_prefix(std::ostream& out) const override {
+    out << '[' << shard << "] ";
+  }
+};
+
+auto sync_shard(const DoutPrefixProvider* dpp,
+                std::string_view period, uint32_t shard,
+                RemoteLog& peer_log, LocalLog& local_log,
+                RemoteMetadata& peer_meta, LocalMetadata& local_meta,
+                LogStatus& log_status)
+    -> awaitable<void>
+{
+  auto prefix = ShardPrefix{*dpp, shard};
+  dpp = &prefix;
+
+  using namespace std::chrono_literals;
+  constexpr ceph::timespan duration = 120s; // TODO
+  std::string cookie = "TODO";
+  auto lock = ShardLockClient{log_status, shard, cookie};
+
+  // retry on lock contention
+  for (;;) {
+    try {
+      co_await with_lease(lock, duration,
+          sync_shard_locked(dpp, period, shard, peer_log, local_log,
+                            peer_meta, local_meta, log_status));
+    } catch (const boost::system::system_error& e) {
```

Comment:
oops, this should also catch the `lease_aborted` exceptions discussed in https://github.com/ceph/ceph/pull/49904#issuecomment-1413872191